prairielib/lib/sql-db.js

```diff
@@ -124,762 +112,809 @@ function escapeIdentifier(identifier) {
   return pg.Client.prototype.escapeIdentifier(identifier);
 }
 
-/**
- * The pool from which clients will be acquired.
- * @type { import("pg").Pool }
- */
-let pool = null;
+class PostgresPool {
+  constructor() {
+    /**
+     * @type {import('pg').Pool}
+     * @private
+     *
+     * The pool from which clients will be acquired.
+     * */
+    this.pool = null;
+
+    /**
+     * @type {AsyncLocalStorage<import('pg').PoolClient>}
+     * @private
+     *
+     * We use this to propagate the client associated with the current transaction
+     * to any nested queries. In the past, we had some nasty bugs associated with
+     * the fact that we tried to acquire new clients inside of transactions, which
+     * ultimately lead to a deadlock.
+     */
+    this.alsClient = new AsyncLocalStorage();
```

Comment:
Do we in fact still want the AsyncLocalStorage code here? I'm worried that we might inadvertently rely on it, only to have it break in some complex and impossible-to-debug way. An alternative might be to use AsyncLocalStorage to throw an error if anyone tries to nest a DB call inside a transaction?

Reply:
Until we merge https://github.com/PrairieLearn/PrairieLearn/pull/5489, I want to keep this here. Even then, I think it's still useful, and it mirrors how transactions work in things like Ruby on Rails (no need to manually pipe around a client, which is a footgun). That said, after the other PR, we really shouldn't have any more transactions, so this code should remain unused.